samples/wifi/provisioning/softap/src/main.c

```diff
@@ -0,0 +1,283 @@
+
+		LOG_PANIC();
+		IF_ENABLED(CONFIG_REBOOT, (sys_reboot(0)));
+		break;
+	case SOFTAP_WIFI_PROVISION_EVT_FATAL_ERROR:
+		LOG_ERR("Provisioning failed, fatal error!");
+		FATAL_ERROR();
+		break;
+	default:
+		/* Don't care */
+		return;
+	}
+}
+
+static void button_handler(uint32_t button_states, uint32_t has_changed)
+{
+	if ((has_changed & DK_BTN1_MSK) && (button_states & DK_BTN1_MSK)) {
+		LOG_INF("Button 1 pressed");
+
+		int ret = softap_wifi_provision_reset();
+
+		if (ret) {
+			LOG_ERR("softap_wifi_provision_reset, error: %d", ret);
+			FATAL_ERROR();
+			return;
+		}
+	}
+}
+
+static int wifi_power_saving_disable(void)
+{
+	int ret;
+	struct net_if *iface = net_if_get_first_wifi();
+		struct wifi_ps_params params = {
+		.enabled = WIFI_PS_DISABLED
+	};
+
+	ret = net_mgmt(NET_REQUEST_WIFI_PS, iface, &params, sizeof(params));
+	if (ret) {
+		LOG_ERR("Failed to disable PSM, error: %d", ret);
+		FATAL_ERROR();
+		return ret;
+	}
+
+	return 0;
+}
+
+static int wifi_power_saving_enable(void)
+{
+	int ret;
+	struct net_if *iface = net_if_get_first_wifi();
+	struct wifi_ps_params params = {
+		.enabled = WIFI_PS_ENABLED
+	};
+
+	ret = net_mgmt(NET_REQUEST_WIFI_PS, iface, &params, sizeof(params));
+	if (ret) {
+		LOG_ERR("Failed to enable PSM, error: %d", ret);
+		FATAL_ERROR();
+		return ret;
+	}
+
+	return 0;
+}
+
+/* Function used to disable and re-enable PSM after a configured amount of time post provisioning.
+ * This is to ensure that the device is discoverable via mDNS so that clients can
+ * confirm that provisioning succeeded. This is needed due to m(DNS) SD being unstable
+ * in Power Save Mode.
+ */
+void psm_set(void)
+{
+	int ret;
+
+	ret = wifi_power_saving_disable();
+	if (ret) {
+		LOG_ERR("wifi_power_saving_disable, error: %d", ret);
+		FATAL_ERROR();
+		return;
+	}
+
+	LOG_INF("PSM disabled");
+
+	k_sleep(K_SECONDS(CONFIG_SOFTAP_WIFI_PROVISION_SAMPLE_PSM_DISABLED_SECONDS));
+
+	ret = wifi_power_saving_enable();
+	if (ret) {
+		LOG_ERR("wifi_power_saving_enable, error: %d", ret);
+		FATAL_ERROR();
+		return;
+	}
+
+	LOG_INF("PSM enabled");
+}
+
+int main(void)
+{
+	int ret;
+	bool provisioning_completed = false;
+
+	LOG_INF("SoftAP Wi-Fi provision sample started");
+
+	ret = dk_buttons_init(button_handler);
+	if (ret) {
+		LOG_ERR("dk_buttons_init, error: %d", ret);
+		FATAL_ERROR();
+		return ret;
+	}
+
+	ret = dk_leds_init();
+	if (ret) {
+		LOG_ERR("dk_leds_init, error: %d", ret);
+		FATAL_ERROR();
+		return ret;
+	}
+
+	ret = softap_wifi_provision_init(softap_wifi_provision_handler);
+	if (ret) {
+		LOG_ERR("softap_wifi_provision_init, error: %d", ret);
+		FATAL_ERROR();
+		return ret;
+	}
+
+	ret = conn_mgr_all_if_up(true);
+	if (ret) {
+		LOG_ERR("conn_mgr_all_if_up, error: %d", ret);
+		FATAL_ERROR();
+		return ret;
+	}
+
+	LOG_INF("Network interface brought up");
+
+	ret = softap_wifi_provision_start();
+	switch (ret) {
+	case 0:
+		provisioning_completed = true;
+		break;
+	case -EALREADY:
```

Comment:
Not really, if the device has already been provisioned after boot (stored credentials). There is no point of disabling psm for a while because there will be no open the wifiprov SD service has been registered too.